### PR TITLE
Allow set sts as call credential for opencensus tracer stub

### PR DESCRIPTION
**Pull request by bianpengyuan**
Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

cc @mandarjog @kyessenov @JimmyCYJ

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by yanavlasov** [comment]
Does this change need unit tests?

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc [hunk 1]
```diff
@@ -28,6 +28,8 @@ namespace Extensions {
 namespace Tracers {
 namespace OpenCensus {
 
+constexpr char kGoogleStackdriverTraceAddress[] = "cloudtrace.googleapis.com";
```
Comment (kyessenov) [inline source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: nit: Envoy style does not use `kConstant`, just `Constant` or `CONSTANT`.

File: api/envoy/config/trace/v2/trace.proto [hunk 2]
```diff
@@ -184,6 +184,13 @@ message OpenCensusConfig {
   // https://github.com/grpc/grpc/blob/master/doc/naming.md
   string stackdriver_address = 10;
 
+  // (optional) The call credentials used in gRPC call to production Stackdriver.
+  // Currently only :ref:`sts_service <envoy_v3_api_field_config.core.v3.GrpcService.GoogleGrpc.CallCredentials.sts_service>`
+  // is supported by opencensus tracer. If not provided, or the provided option is not
```
Comment (kyessenov) [inline api/envoy/config/trace/v2/trace.proto]: From the code it looks like default credentials are used in addition to STS. Is that what gRPC normally does?

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc [hunk 3]
```diff
@@ -251,6 +270,14 @@ Driver::Driver(const envoy::config::trace::v3::OpenCensusConfig& oc_config,
...
           grpc::CreateChannel(oc_config.stackdriver_address(), grpc::InsecureChannelCredentials());
       opts.trace_service_stub = ::google::devtools::cloudtrace::v2::TraceService::NewStub(channel);
+    } else if (oc_config.has_stackdriver_call_credentials() &&
+               oc_config.stackdriver_call_credentials().has_sts_service()) {
+      // If call credential is provided and is STS service, create a stub with that credential.
+      auto call_creds = getStackdriverCallCredential(oc_config);
+      auto channel_creds =
+          grpc::CompositeChannelCredentials(grpc::GoogleDefaultCredentials(), call_creds);
+      auto channel = grpc::CreateChannel(kGoogleStackdriverTraceAddress, channel_creds);
+      opts.trace_service_stub = ::google::devtools::cloudtrace::v2::TraceService::NewStub(channel);
```
Comment (kyessenov) [inline source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: I think we need a unit test that covers these lines. Somewhere in https://github.com/envoyproxy/envoy/blob/253bff42ee36da6b2190c50e1e9984454d453bb7/test/extensions/tracers/opencensus/config_test.cc. I don't think it's possible to test STS end-to-end, or how to introspect gRPC stubs, but at least we should verify that the credentials if config says so.

File: api/envoy/config/trace/v2/trace.proto (hunk 2 above)
Comment (bianpengyuan) [reply api/envoy/config/trace/v2/trace.proto]: Yeah oc library uses default credential for both channel and call credentials. https://github.com/census-instrumentation/opencensus-cpp/blob/c900c4d723ef596f357b0e695e68e2fa725eec90/opencensus/exporters/stats/stackdriver/internal/stackdriver_exporter.cc#L91

This change make it use sts for call credentials but still uses google default for channel credentials.

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc (hunk 1 above)
Comment (bianpengyuan) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: Done.

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc (hunk 3 above)
Comment (bianpengyuan) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: Added a unit test to verify that a tracer could be constructed with sts configured.

**Comment by kyessenov** [comment]
The coverage build failure is a tough nut to crack. I had most success with clang-asan coverage build and then fixing it ODR violations between tests.

File: api/envoy/config/trace/v2/trace.proto [hunk 4]
```diff
@@ -184,6 +184,13 @@ message OpenCensusConfig {
...
+  // :ref:`sts_service <envoy_v3_api_field_config.core.v3.GrpcService.GoogleGrpc.CallCredentials.sts_service>`,
+  // the default Google credential will be used as call credential.
+  api.v2.core.GrpcService.GoogleGrpc.CallCredentials stackdriver_call_credentials = 13;
```
Comment (lizan) [inline api/envoy/config/trace/v2/trace.proto]: Shouldn't we just have an `api.v2.core.GrpcService` service to cover where OpenCensus connect to and the options?

Comment (bianpengyuan) [reply api/envoy/config/trace/v2/trace.proto]: Opencensus library uses its own network stack, not sure how can I use envoy google gRPC stack to replace that. It provides an mechanism to specify a gRPC stub, but it expects stub of a specific proto (https://github.com/census-instrumentation/opencensus-cpp/blob/c900c4d723ef596f357b0e695e68e2fa725eec90/opencensus/exporters/trace/stackdriver/stackdriver_exporter.h#L40). Will GoogleStub be compatible with that?

On the other hand, Opencensus library is unlikely to utilize all the options provided by the GrpcService. I can only foresee that channel credential might be needed in the future. So adding the whole gRPC service here seems an overkill for me.

**Comment by bianpengyuan** [comment]
The test will crash on non-Google instance. `CompositeChannelCredentials` expect channel credential not to be null.

File: api/envoy/config/trace/v2/trace.proto (hunk 4 above)
Comment (lizan) [reply api/envoy/config/trace/v2/trace.proto]: That is a generated Google gRPC stub from the code? So, you should be able to reuse (might need some refactor) from [here](https://github.com/envoyproxy/envoy/blob/5248a4fb7d4c2a3d1fa151f944d3a63f6b7a06cf/source/common/grpc/google_async_client_impl.cc#L83-L85), i.e. calling `google::devtools::cloudtrace::v2::TraceService::NewStub` on the channel created by these code. Then you don't have to worry about other features added in gRPC not being picked up by OpenCensus.

You can add "only Google gRPC in GrpcService is supported" doc (and validate in code) so `EnvoyGrpc` won't be configured.

Comment (bianpengyuan) [reply api/envoy/config/trace/v2/trace.proto]: Done.

**Comment by bianpengyuan** [comment]
@lizan Done.

File: docs/root/intro/version_history.rst [hunk 5]
```diff
@@ -19,7 +19,8 @@ Version history
   "envoy.reloadable_features.new_http2_connection_pool_behavior" and then re-configure your clusters or
   restart Envoy. The behavior will not switch until the connection pools are recreated. The new
   circuit breaker behavior is described :ref:`here <arch_overview_circuit_break>`.
-* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`.
+* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`. 
+* tracing: added gRPC service configuration to the OpenCensus Stackdriver tracer.
```
Comment (lizan) [inline docs/root/intro/version_history.rst]: nit: alphabetical order

Comment (bianpengyuan) [reply docs/root/intro/version_history.rst]: Somehow I though `t` is after `u` :)

**Pull request merged**